lec13-time-series.Rmd

```diff
@@ -0,0 +1,940 @@
+Let's start with $(0,0)$: both derivatives are 0, so there's no arrow. For $(0,1)$, $dx/dt = 1$
+and $dv/dt = 0$, so we draw an arrow pointing horizontally in the $+x$ direction. Similarly for
+$(1,0)$. For $(1,1)$, we draw an arrow that goes $1$ unit in the $x$ direction and $-1$ unit in
+the $y$ direction. As we keep filling in this grid, we start to get a picture of how the system
+can behave for different starting points of $x$ and $v$. 
+
+This is analogous to plotting $dN/dt$ vs. $N$ and drawing arrows to the right where $dN/dt$ is
+positive and arrows to the left when it's negative, except that now we have to consider that 
+the rate of change is affecting both variables at the same time. 
+
+#### Challenge
+
+Define a function called `mass_spring` that uses the format required by `ode` to calculate 
+and return $dx/dt$ and $dv/dt$. We will use this function later to plot a phase portrait in R.
+Remember that the arguments for your function must be called `t`, `y`, and `parameters`. 
+
+```{r, eval=FALSE, echo=FALSE}
+# Challenge solution
+
+mass_spring <- function(t, y, parameters) {
+  # calculates dx/dt and dy/dt for a predator-prey model
+  
+  # t: time at which to evaluate derivatives
+  # y: vector of system variables (c(x, v))
+  # parameters: vector of model parameters (c(omega))
+    
+  # the state vector y has two elements because we have two variables
+  x <- y[1] # position
+  v <- y[2] # velocity
+  
+  omega <- parameters[1]
+    
+  # calculate rate of change
+  dx <- v
+  dv <- -omega^2 * x
+  
+  # return rate of change
+  return(list(c(dx, dv)))
+}
+```
+
+### Using `phaseR` to draw phase portraits in R
+
+`phaseR` is a package specifically for drawing phase portraits. It can automatically calculate 
+and plot things like trajectories and vector fields. 
+
+```{r}
+library(phaseR)
+```
+
+```{r, eval=FALSE}
+?phaseR
+```
+
+```{r, eval=FALSE}
+?flowField # the nice thing about this packages is that it takes functions in the same format as `ode`
+```
+
+Let's plot a vector field and some trajectories for the predator-prey model.
+
+```{r}
+# plot vector field: on a grid of points, plot an arrow in the direction of dx/dt and dy/dt
+pp_flowField <- flowField(predator_prey, x.lim = c(0, 10), y.lim = c(0, 10),
+                          parameters = c(a = 5, b = 1, c = 1, d = 0.2), # same parameters as before,
+                          points = 15, # this is the density of grid points on which to plot arrows
+                          system = 'two.dim', # 'two.dim' is default
+                          add = FALSE)
+
+# add trajectories
+pp_trajectory <- trajectory(predator_prey, 
+                            # y0 is a matrix where each row is pairs of (X, Y) 
+                            y0 = matrix(c(1, 1, 8, 5, 2, 0, 0, 8), ncol = 2,
+                                   byrow = TRUE), 
+                            t.end = 20, # how far in time to calculate the trajectories
```

Comment:
`tlim = c(0, 20)` - this was breaking a lecture from last year (`lec06` iirc) that I tested and fixed back in August

Reply:
Fixed for last year's lectures in PR #203 - I realize now I should have done this in a separate PR...